src/main.cpp

```diff
@@ -0,0 +1,416 @@
+#include <iostream>
+#include <map>
+#include <string>
+#include <optional>
+#include <vector>
+#include <memory>
+#include <sstream>
+#include <iterator>
+
+
+
+// ----------------
+// State class
+// ----------------
+
+class State {
+public:
+    class Env {
+        void set_var(const std::string& var_name, const std::string& var_value) {
+            vars_[var_name] = var_value;
+        }
+
+        std::optional<std::string> get_var(const std::string& var_name) const { 
+            const auto it = vars_.find(var_name);
+            
+            if (it != vars_.end()) {
+                return it->second;
+            }
+
+            return {};
+
+        }
+
+    private:
+        std::map<std::string, std::string> vars_;
+    };
+
+};
+
+
+
+// ----------------
+// Streams classes
+// ----------------
+
+class IStreams {
+public:
+    virtual std::istream& get_in_stream() = 0;
+    virtual std::ostream& get_out_stream() = 0;
+    virtual std::ostream& get_err_stream() = 0;
+
+    virtual ~IStreams() = default;
+};
+
+class StreamsGlobal : public IStreams {
+public:
+    std::istream& get_in_stream() override {return std::cin; }
+    std::ostream& get_out_stream() override {return std::cout; }
+    std::ostream& get_err_stream() override {return std::cerr; }
+};
+
+
+
+// ----------------
+// Cmd classes
+// ----------------
+
+class Args {
+public:
+    using ArgType = std::string;
+
+    Args() = default;
+    explicit Args(ArgType arg) : args_{std::move(arg)} {}
+    explicit Args(std::vector<ArgType> args) : args_{std::move(args)} {}
+
+
+    const ArgType& get_arg(size_t i) const {
+         return args_.at(i); 
+    }
+
+    void add_arg(ArgType arg) {
+        args_.push_back(std::move(arg));
+    }
+
+    const std::vector<ArgType>& get_all() const { 
+        return args_; 
+    }
+
+private:
+    std::vector<ArgType> args_;
+};
+
+
+class Cmd {
+public:
+
+    Cmd(std::string cmd, std::vector<std::string> args = {})
+        : cmd_{std::move(cmd)}, args_{std::move(args)} 
+    {
+        // do nothing
+    }
+
+    std::string& get_cmd() { return cmd_; }
+    Args& get_args() { return args_; }
+
+private:
+    std::string cmd_;
+    Args args_;
+};
+
+
+class CmdRaw : public Cmd {
+public:
+    CmdRaw(std::string cmd, std::vector<std::string> args = {})
+        : Cmd(std::move(cmd), std::move(args)) 
+    {
+
+    }
+
+    // substitution will be here
+    void preprocess([[maybe_unused]] State& state) {}
+};
+
+
+
+
+// ----------------
+// CmdParser classes
+// ----------------
+
+class ICmdParser {
```

Comment:
Do we really need virtual Parser?

Reply:
Virtual Parser allows to implement different parsing policies (how process quotes, pipes, streams, e.t.c) without changing the shell class (there will be only ptr to ICmdParser )

Reply:
It seems to me like overengineering, because there is no requirement of such behavior.
But we can keep this virtual class as temporary solution.